gcc_select/c99.1
.\"
.\" Copyright (c) 1997 Joerg Wunsch
.\"
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE DEVELOPERS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE DEVELOPERS BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" FreeBSD: src/usr.bin/c89/c89.1,v 1.4.2.4 2001/08/02 01:11:13 obrien Exp
.\" $FreeBSD: src/usr.bin/c99/c99.1,v 1.3 2004/03/04 20:29:27 josef Exp $
.\"
.Dd October 7, 2002
.Os
.Dt C99 1
.Sh NAME
.Nm c99
.Nd standard C language compiler
.Sh SYNOPSIS
.Nm
.Op Fl cEgs
.Op Fl D Ar name Ns Op = Ns Ar value
.Ar ...
.Op Fl I Ar directory ...
.Op Fl L Ar directory ...
.Op Fl o Ar outfile
.Op Fl O Ar optlevel
.Op Fl U Ar name ...
.Ar operand ...
.Sh DESCRIPTION
This is the name of the C language compiler as required by the
.St -p1003.1-2001
standard.
.Pp
The
.Nm
compiler accepts the following options:
.Bl -tag -width indent
.It Fl c
Suppress the link-edit phase of the compilation, and do not remove any
object files that are produced.
.It Fl D Ar name Ns Op = Ns Ar value
Define name as if by a C-language
.Ic #define
directive.
If no
.Dq = Ns Ar value
is given, a value of 1 will be used.
Note that in order to request a
translation as specified by
.St -p1003.1-2001 ,
you need to define
.Dv _POSIX_C_SOURCE=200112L
either in the source or using this option.
The
.Fl D
option has lower precedence than the
.Fl U
option.
That is, if
.Ar name
is used in both a
.Fl U
and a
.Fl D
option,
.Ar name
will be undefined regardless of the order of the options.
The
.Fl D
option may be specified more than once.
.It Fl E
Copy C-language source files to the standard output, expanding all
preprocessor directives; no compilation will be performed.
.It Fl g
Produce symbolic information in the object or executable files.
.It Fl I Ar directory
Change the algorithm for searching for headers whose names are not
absolute pathnames to look in the directory named by the
.Ar directory
pathname before looking in the usual places.
Thus, headers whose
names are enclosed in double-quotes
.Pq Qq
will be searched for first
in the directory of the file with the
.Ic #include
line, then in
directories named in
.Fl I
options, and last in the usual places.
For headers whose names are enclosed in angle brackets
.Pq Aq ,
the header
will be searched for only in directories named in
.Fl I
options and then in the usual places.
Directories named in
.Fl I
options shall be searched in the order specified.
The
.Fl I
option may be specified more than once.
.It Fl L Ar directory
Change the algorithm of searching for the libraries named in the
.Fl l
objects to look in the directory named by the
.Ar directory
pathname before looking in the usual places.
Directories named in
.Fl L
options will be searched in the order specified.
The
.Fl L
option may be specified more than once.
.It Fl o Ar outfile
Use the pathname
.Ar outfile ,
instead of the default
.Pa a.out ,
for the executable file produced.
.It Fl O Ar optlevel
If
.Ar optlevel
is zero, disable all optimizations.
Otherwise, enable optimizations at the specified level.
.It Fl s
Produce object and/or executable files from which symbolic and other
information not required for proper execution has been removed
(stripped).
.It Fl U Ar name
Remove any initial definition of
.Ar name .
The
.Fl U
option may be specified more than once.
.It Fl W Cm 64
Make pointers 64 bits.  There is presently no
.Fl W Cm 32
option, since it is the default.
.El
.Pp
An operand is either in the form of a pathname or the form
.Fl l
library.
At least one operand of the pathname form needs to be specified.
Supported operands are of the form:
.Bl -tag -offset indent -width ".Fl l Ar library"
.It Ar file Ns Pa .c
A C-language source file to be compiled and optionally linked.
The operand must be of this form if the
.Fl c
option is used.
.It Ar file Ns Pa .a
A library of object files, as produced by
.Xr ar 1 ,
passed directly to the link editor.
.It Ar file Ns Pa .o
An object file produced by
.Nm Fl c ,
and passed directly to the link editor.
.It Fl l Ar library
Search the library named
.Pa lib Ns Ar library Ns Pa .a .
A library will be searched when its name is encountered, so the
placement of a
.Fl l
operand is significant.
.El
.Sh SEE ALSO
.Xr ar 1 ,
.Xr c89 1 ,
.Xr cc 1
.Sh STANDARDS
The
.Nm
utility interface conforms to
.St -p1003.1-2001 .


X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


gcc_select/c99.c
/*-
 * Copyright (c) 2002 Tim J. Robbins.
 * Copyright (c) 2007 Apple Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * c99 -- compile standard C programs
 *
 * This is essentially a wrapper around the system C compiler that forces
 * the compiler into C99 mode and handles some of the standard libraries
 * specially.
 */

#include <sys/cdefs.h>
/* __FBSDID("$FreeBSD: src/usr.bin/c99/c99.c,v 1.2 2002/10/08 02:19:54 tjr Exp $");*/

#include <sys/types.h>

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sysexits.h>

char **args;
u_int cargs, nargs;

void addarg(const char *);
void addlib(const char *);
void combine_and_addarg (const char *, const char *);
void usage(void);

/* Store -U args.  */
char **undef_args;
u_int undef_cargs, undef_nargs;
void record_undef(const char *);

void add_def(const char *);
void combine_and_addarg (const char *, const char *);

int
main(int argc, char *argv[])
{
	int link = 1;
	int inputs = 0;
	int verbose = 0;

	addarg("cc");
	addarg("-std=iso9899:1999");
	addarg("-pedantic");
	addarg("-Wextra-tokens"); /* Radar 4205857 */
#if defined (__ppc__) || defined (__ppc64__)
	/* on ppc long double doesn't work */
	addarg("-mlong-double-64");
#endif
	addarg("-fmath-errno");  /* Radar 4011622 */
	addarg("-fno-builtin-pow");
	addarg("-fno-builtin-powl");
	addarg("-fno-builtin-powf");

	for (;;)
	  {
	    int ch = getopt(argc, argv, "cD:EgI:L:o:O:sU:W:l:");
	    if (optind >= argc && ch == -1)
	      break;
	    switch (ch)
	      {
	      case 'c':
		addarg ("-c");
		link = 0;
		break;
	      case 'D':
		add_def (optarg);
		break;
	      case 'E':
		addarg ("-E");
		link = 0;
		break;
	      case 'g':
		addarg ("-g");
		break;
	      case 'I':
		combine_and_addarg ("-I", optarg);
		break;
	      case 'L':
		combine_and_addarg ("-L", optarg);
		break;
	      case 'o':
		addarg ("-o");
		addarg (optarg);
		break;
	      case 'O':
		combine_and_addarg ("-O", optarg);
		break;
	      case 's':
		addarg ("-s");
		break;
	      case 'U':
		record_undef (optarg);
		combine_and_addarg ("-U", optarg);
		break;
	      case 'W':
		if (strcmp (optarg, "64") == 0)
		  addarg ("-m64");
		else if (strcmp (optarg, "verbose") == 0)
		  {
		    addarg ("-v");
		    verbose = 1;
		  }
		else
		  errx(EX_USAGE, "invalid argument `%s' to -W", optarg);
		break;
	      case 'l':
		addlib (optarg);
		break;
	      case -1:
		if (strcmp (argv[optind-1], "--") == 0)
		  {
		    while (optind < argc)
		      {
			if (argv[optind][0] == '-')
			  combine_and_addarg ("./", argv[optind]);
			else
			  addarg (argv[optind]);
			inputs++;
			optind++;
		      }
		  }
		else
		  {
		    addarg (argv[optind++]);
		    inputs++;
		  }
		break;
	      case '?':
		usage ();
		break;
	      }
	  }

	if (link && inputs > 0) {
	  addarg("-liconv");
	}
	if (verbose)
	  {
	    int i;
	    for (i = 0; args[i]; i++)
	      printf ("\"%s\" ", args[i]);
	    putchar ('\n');
	  }
	execv("/usr/bin/cc", args);
	err(EX_OSERR, "/usr/bin/cc");
}

/* Combine item1 and item2 and used them as one argument.
   This is used to combine "-O" with "1" to make "-O1".  */
void
combine_and_addarg (const char *item1, const char *item2)
{
  char *item = (char *) malloc (sizeof (char) * (strlen(item1) + strlen(item2) + 1));
  if (item == NULL)
    err (EX_OSERR, "malloc");
  strcpy (item, item1);
  strcat (item, item2);
  addarg (item);
}

/* Record macro undefs on the command line. -U. */
void 
record_undef (const char *item)
{
  if (undef_nargs + 1 > undef_cargs) {
    undef_cargs += 16;
    undef_args = realloc(undef_args, sizeof(*undef_args) * undef_cargs);
    if (undef_args == NULL)
      err(EX_OSERR, "malloc");
  }

  if ((undef_args[undef_nargs++] = strdup(item)) == NULL)
    err(EX_OSERR, "strdup");
  undef_args[undef_nargs] = NULL;
}

/* Add command line macro def (-Dblah) in the argument list.
   If it is was undefined earlier than do not add.  Radar 3762036.  */
void
add_def (const char *item)
{
  u_int i;
  const char * equ_pos;
  
  equ_pos = strchr (item, '=');
  if (equ_pos == NULL)
    equ_pos = item + strlen (item);

  for (i = 0; i < undef_nargs; i++)
    if (strncmp (item, undef_args[i], equ_pos - item) == 0
	&& undef_args[i][equ_pos - item] == '\0')
      return;
  combine_and_addarg ("-D", item);
}

void
addarg(const char *item)
{
  if (nargs + 2 > cargs) {
    cargs += 16;
    if ((args = realloc(args, sizeof(*args) * cargs)) == NULL)
      err(EX_OSERR, "malloc");
  }
  
  args[nargs++] = strdup (item);
  
  if (args[nargs-1] == NULL)
    err(EX_OSERR, "strdup");
  args[nargs] = NULL;
}

void
addlib(const char *lib)
{
  if (strcmp(lib, "pthread") == 0)
    /* pthread functionality is in libc. */
    return;
  if (strcmp(lib, "rt") == 0)
    /* librt functionality is in libc or unimplemented. */
    return;
  if (strcmp(lib, "xnet") == 0)
    /* xnet functionality is in libc. */
    return;

  combine_and_addarg("-l", lib);
}

void
usage(void)
{
	fprintf(stderr,
"usage: c99 [-cEgs] [-D name[=value]] [-I directory] ... [-L directory] ...\n");
	fprintf(stderr,
"       [-o outfile] [-O optlevel] [-U name]... [-W 64] operand ...\n");
	exit(1);
}
